subprojects/data-structures/src/main/java/org/gradle/internal/collect/Nil.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.collect;
+
+import java.util.function.Consumer;
+
+class Nil<T> implements PersistentList<T> {
```

Comment:
`Nil` shouldn't have a type parameter. instead, there should be a static `of()` method that casts the singleton `Nil` instance to `PersistentList<T>`.

Reply:
There is now no extra `Nil` class, only a `NIL` constant of type `PersistentList<Object>`.